### ISSUE TITLE
[Bug?] `jax.random.shuffle` does not behave like `np.random.shuffle` in terms of shuffling along a given axis

### ISSUE THREAD

**Issue by zmjjmz** [google/jax#2066]
When we call `jax.random.shuffle(key, array, axis=0)` on an ndarray of shape (n,m) I'm expecting (based on the docs) that we'll end up with an new array where every _row_ has the same order internally but the order of the rows is shuffled. I'm not super clear that this is guaranteed by the [docs](https://jax.readthedocs.io/en/latest/jax.random.html#jax.random.shuffle), which say:

>Shuffle the elements of an array uniformly at random along an axis.

But the numpy docs for [`numpy.random.shuffle`](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.shuffle.html) do make this super clear, saying that 

>This function only shuffles the array along the first axis of a multi-dimensional array. The order of sub-arrays is changed but their contents remains the same.

So I'm not super sure the _intent_ is for the Jax behavior to mimic the numpy behavior here, but that is what I was expecting. To demonstrate / reproduce this difference in behavior, see this [gist](https://gist.github.com/zmjjmz/099afcefe9ea3cf3e8015dc8c2e7a340)

**Comment by ChrisWaites** [comment]
+1

**Comment by jakevdp** [comment]
I'm not certain whether the intent was to duplicate numpy, but I do think it's confusing that this function does not. I propose the following sequence to address this, while avoiding silent breakages of existing code.

- move the ``jax.random.shuffle`` to ``jax.random.shuffle_index``
- point ``jax.random.shuffle`` to the new function with an informative deprecation warning
- after a couple releases, change ``jax.random.shuffle`` to match the semantics of ``np.random.shuffle``

**Comment by jekbradbury** [comment]
This definitely seems like a footgun. As you suggest, the implemented semantics are quite different from `np.random.shuffle`: they shuffle _elements_ along the specified axis (with a different permutation for each point in the rank-N-minus-1 subarray) rather than shuffling subarrays along that axis. This discrepancy dates back to the original open source release of JAX.

One reason this didn't come up earlier is that the large majority of uses of `jax.random.shuffle` in the wild pass a rank-1 array, where the two behaviors are the same. In fact, adding an assertion that the array is rank 1 passes first-line internal tests, so there may be as few as _zero_ internal users depending on the element-shuffling aspect of the current semantics. That makes me think we might even be able to get away with changing the behavior without a full deprecation process...

**Comment by mattjj** [comment]
@jakevdp that plan SGTM.

**Comment by jekbradbury** [comment]
@jakevdp despite my willingness to jump the gun and switch the behavior without deprecation, the best course of action is probably to follow your proposed steps (while omitting the deprecation warning for rank-1 inputs). Are you interested in taking that on?

**Comment by jakevdp** [comment]
Sure, I'll take it

**Comment by shoyer** [comment]
Do we even need `random.shuffle` at all? The semantics of the NumPy function are to do in-place modification, so in practice I never use it. Explicit indexing and permutation + vmap seem like good alternatives.

**Comment by jakevdp** [comment]
> Do we even need random.shuffle at all?

Agreed  – now that I look at it, shuffle and permutation have essentially identical behavior for array inputs, except shuffle is in-place while permutation returns a copy. With that in mind, a cleaner solution would be to deprecate `jax.random.shuffle`, and only provide `jax.random.permutation` (possibly leaving a stub for ``shuffle`` with a ``NotImplementedError`` and a pointer to ``permutation``).

**Comment by jakevdp** [comment]
Fixed in #2925 

`shuffle()` is now deprecated in favor of `permutation()`, which has the same axis shuffling semantics as numpy.
